cache/cache_rep.cc

```diff
@@ -0,0 +1,35 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+//
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+#include <cassert>
+#include <cstddef>
+#include <cstring>
+#include <memory>
+
+#include "cache/cache_rep.h"
+#include "rocksdb/cache.h"
+#include "rocksdb/slice.h"
+#include "table/block_based/block_based_table_reader.h"
+#include "util/coding.h"
+
+namespace ROCKSDB_NAMESPACE {
+CacheRep::CacheRep(std::shared_ptr<Cache> cache)
+    : cache_allocated_size_(0) {
+  assert(cache != nullptr);
+  cache_ = cache;
+  std::memset(cache_key_, 0, kCacheKeyPrefixSize + kMaxVarint64Length);
```

Comment:
Also may not be necessary.

Reply:
Good question! I was wondering the same thing when developing :)

However, I think L30 is still needed mainly for the following reason:
L32 in the function "GetNextCacheKey()" actually stores the encoded value of "next_cache_key_id_" in-place into "cache_key_" as the non-prefix part of the cache key. In another word, we generate a new non-prefix part of the cache key every time but keep storing it in the same place (the place is "cache_key_ + kCacheKeyPrefixSize").

Therefore, we would like to reset the place "cache_key_ + kCacheKeyPrefixSize" every time when we want to store a new non-prefix part of the cache key.  

The reason for why we would like to store cache key in place, I guess, is for memory and speed efficiency. 

But now reading your question make me consider it might be worthwhile to add a comment for line 30 ... let me see ...

Reply:
`[cache_key_, cache_key_ + kCacheKeyPrefixSize)` doesn't change, and `EncodeVarint64()` overwrites `[cache_key_ + kCacheKeyPrefixSize, end)`. Then the `Slice` covers `[cache_key_, end)` which feels safe, though the bytes beyond `end` could be nonzero and stale.